src/Mailgun/Mailgun.php

```diff
@@ -36,17 +43,79 @@ class Mailgun
     protected $apiKey;
 
     /**
-     * @param string|null $apiKey
-     * @param HttpClient  $httpClient
-     * @param string      $apiEndpoint
+     * @var HttpMethodsClient
+     */
+    private $httpClient;
+
+    /**
+     * @var ResponseSerializer
+     */
+    private $serializer;
+
+    /**
+     * @param string|null             $apiKey
+     * @param HttpClient|null         $httpClient
+     * @param string                  $apiEndpoint
+     * @param ResponseSerializer|null $serializer
      */
     public function __construct(
         $apiKey = null,
         HttpClient $httpClient = null,
-        $apiEndpoint = 'api.mailgun.net'
+        $apiEndpoint = 'api.mailgun.net',
+        ResponseSerializer $serializer = null
     ) {
         $this->apiKey = $apiKey;
+        $this->serializer = $serializer ?: new ObjectSerializer();
         $this->restClient = new RestClient($apiKey, $apiEndpoint, $httpClient);
+
+        $plugins = [
+            new Plugin\AddHostPlugin(UriFactoryDiscovery::find()->createUri('https://api.mailgun.net')),
+            new Plugin\HeaderDefaultsPlugin([
+                'User-Agent' => 'mailgun-sdk-php/v2 (https://github.com/mailgun/mailgun-php)',
+                'Authorization' => 'Basic '.base64_encode(sprintf('api:%s', $apiKey)),
+            ]),
+        ];
+
+        $this->httpClient = new HttpMethodsClient(
+            new PluginClient($httpClient ?: HttpClientDiscovery::find(), $plugins),
+            MessageFactoryDiscovery::find()
+        );
+    }
+
+    /**
+     * @param string $name
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return mixed
+     */
+    public function api($name)
```

Comment:
Cleaner if you know what's in the API and know how the magic works. But for someone who just checks the code it might not be.

Not to mention that from a testability point of view you still have to test all the possible methods, which also does not seem to be cleaner for me: you test non existing methods?

I guess brevity is an argument here, but I feel there are more downsides, while the downside of having many public members is "just" maintenance.